rfcs/0008-serialization/0008-serialization.md

```diff
@@ -2,22 +2,255 @@
+| struct |                                                  | filed-0 | field-1 | ... | field-N |
+| fixvec | items-count                                      |  item-0 |  item-1 | ... |  item-N |
+| dynvec | full-size | offset-0 | offset-1 | ... | offset-N |  item-0 |  item-1 | ... |  item-N |
+| table  | full-size | offset-0 | offset-1 | ... | offset-N | filed-0 | field-1 | ... | field-N |
+| option |                                                  | item or none (zero bytes)         |
+| union  | item-type-id                                     | item                              |
+```
+
+- All items in Header is 32 bit unsigned integers in little-endian
+
+### Primitive Type
+
+#### `byte`
+
+The `byte` is a byte.
+
+##### Examples
+
+`00` is a `byte`.
+
+### Composite Types
+
+### `array`
+
+The `array` is a fixed-size type: it has a fixed-size inner type and a fixed length.
+The size of an `array` is the size of inner type times the length.
+
+Serialize an `array` only need to serialize all items in it. No extra cost for `array` itself.
+
+Each item in an `array` are close to the previous item and the next item.
+There are no extra space between two adjacent items.
+
+#### Examples
+
+If we define `array Byte3 [byte; 3];`, and we want to store three bytes: first is `01`, the second is `02` and the last is `03`, then the serialized bytes will be `01 02 03`.
+
+If we define `array Uint32 [byte; 4];` , and we want to store a 32 bit unsigned integer `0x01020304` into it in little-endian, then the serialized bytes will be `04 03 02 01`.
+
+If we define `array TwoUint32 [Uint32; 4];`, and we want to store two 32 bit unsigned integers in little-endian: first is `0x01020304` and second is `0xabcde`, then the serialized bytes will be `04 03 02 01 de bc 0a 00`.
+
+### `struct`
+
+The `struct` is a fixed-size type: all fields in `struct` is fixed-size and it has a fixed quantity of fields.
+The size of a `struct` is the sum of all fields' size.
+
+Serialize a `struct` only need to serialize all fields in it. No extra cost for `struct` itself.
+
+Fields in an `struct`  are stored in the order they are declared.
+Each fields in an `struct` are close to the previous field and the next field.
+There are no extra space between two adjacent fields.
+
+#### Examples
+
+If we define `struct OnlyAByte { f1: byte }`, and we want to store a byte `ab`, then the serialized bytes will be `ab`.
+
+If we define `struct ByteAndUint32 { f1: byte, f2: Uint32 }`, and we want to store a byte `ab` and a 32 bit unsigned integer `0x010203` in little-endian, then the serialized bytes will be `ab 03 02 01 00`.
+
+### vectors
+
+There two kinds of vectors: fixed vector `fixvec` and dynamic vector `dynvec`.
+
+A vector is fixed or dynamic that depends on the type of it's inner item: if the inner item is fixed-size, then it's a `fixvec`; if the inner item is dynamic-size, then it's a `dynvec`.
```

Comment:
```suggestion
Whether a vector is fixed or dynamic depends on the type of its inner item: if the inner item is fixed-size, then it's a `fixvec`; otherwise, it's a `dynvec`.
```